### PR TITLE
Field labels in activity log

### PR DESCRIPTION
**Pull request by rmunn**
Now when a lexical entry is updated, the activity log will contain not just the differences, but also the human-readable field labels for those differences. This will enable the activity log frontend to generate a complete activity-log entry without needing a second server round-trip.

---
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/265)

### CODE REVIEW

**Comment by irahopkinson** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php, line 45 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJrijF7H5N70xOW8-b:-LDJrijF7H5N70xOW8-c:br064z6) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php#L45)):*
> ```PHP
>         $currentList = $fieldList;
>         $result = '';
>         foreach ($parts as $part) {
> ```

It looks like even though it loops through all `$parts`, the `$result` will be just from the last part . The tests seem to confirm that ts working but I'm finding this hard to follow. What did I miss something? 

---

*[src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php, line 48 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJotzSAJP91fhAX2-t:-LDJotzSAJP91fhAX2-u:b4skawv) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php#L48)):*
> ```
> $fieldId
> ```
Everywhere else in code this should be referred to as `$fieldName`. Fixing this would help with consistency.

---

*[src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php, line 52 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJn_hzCxOLlFDQcd5r:-LDJn_hzCxOLlFDQcd5s:b5d9k7l) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php#L52)):*
> ```PHP
>                 $fieldConfig = $currentList->fields[$fieldId];
>                 $result = $fieldConfig->label;
>                 if ($fieldConfig instanceof LexConfigFieldList) {
> ```

You could just check if `$fieldConfig->type == LexConfig::FIELDLIST` instead of using `instanceof`.

---

*[src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php, line 56 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJoFVxD6IShRuuYTG2:-LDJoFVxD6IShRuuYTG3:bqjxjcp) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php#L56)):*
> ```PHP
>                 }
>             } else {
>                 // If we can't find a label, use the field ID as that's better than nothing
> ```

Hmm, if the data isn't migrated it will show **senses** instead of **Meaning**. **examples** instead of **Example** isn't so bad but not ideal.

---

*[src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php, line 74 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJjius3Gc0xAgwQtcE:-LDJjius3Gc0xAgwQtcF:b472ejk) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php#L74)):*
> ```PHP
>         $result = $differences;
>         foreach ($differences as $key => $value) {
>             // Key will look like "newValue.senses#482f60da-b32a-45b9-9450-ee8f24791557.definition.en"
> ```

I know this change is extending existing design but using a multi-part key to store what is effectively hierarchical data in a flat structure seems hacky to me. Perhaps access functions to deal with multi-part keys would be a start to ensure we are dealing with them consistently and to hide/separate the implementation?

---

*[src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php, line 76 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJpCFlA3mD-RfL9GJ0:-LDJpCFlA3mD-RfL9GJ1:bl0n2t9) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php#L76)):*
> ```
> if (empty($parts)) continue;
> ```
You used `{ ... }` below for a single statement with `if` but not here. I think its safer to always use them but at least be consistent.

---

*[src/Api/Model/Languageforge/Lexicon/Config/LexConfiguration.php, line 93 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJj1K7C5v-QkMcMteb:-LDJj1K9ADLsyS9y-3Jo:b-2n71lp) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Config/LexConfiguration.php#L93)):*
> ```PHP
> 
>         $this->entry->fields[LexConfig::SENSES_LIST] = new LexConfigFieldList();
>         $this->entry->fields[LexConfig::SENSES_LIST]->label = 'Meaning';
> ```

This doesn't work on existing projects because its only used as a template at project creation. This will require a data migration.

---

*[test/php/model/languageforge/lexicon/command/LexEntryCommandsTest.php, line 220 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJuAXT3qiaQ4lebyV-:-LDJuAXT3qiaQ4lebyV0:bie3xgi) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/test/php/model/languageforge/lexicon/command/LexEntryCommandsTest.php#L220)):*
> ```PHP
>         $differences = $updatedEntry->calculateDifferences($secondUpdatedEntry);
>         $this->assertEquals(['oldValue.lexeme.th' => 'first edit', 'newValue.lexeme.th' => 'second edit'], $differences);
>         $withLabels = LexEntryCommands::addFieldLabelsToDifferences($project->config, $differences);
> ```

Checking that `LexEntryCommands::addFieldLabelsToDifferences` works once is enough. This second test doesn't check anything new. Here and below as appropriate.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/265)*

**Comment by irahopkinson** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php, line 45 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJrijF7H5N70xOW8-b:-LDJyJH2EcsD_1zMPvVs:b-dpz56w) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php#L45)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

It looks like even though it loops through all `$parts`, the `$result` will be just from the last part . The tests seem to confirm that ts working but I'm finding this hard to follow. What did I miss something? 
</blockquote></details>

Ok, I had to step through the code to see how it works.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/265)*

**Comment by rmunn** [comment]
Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php, line 56 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJoFVxD6IShRuuYTG2:-LD_zI9QCEOrKHr3kimg:bc44oc2) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php#L56)):*
<details><summary><i>Previously, rmunn (Robin Munn) wrote…</i></summary><blockquote>

Hmm, yes, the changes I made to LexConfiguration wouldn't be enough. I wonder if it would be a good idea to put in a temporary fix where I hardcode those two strings here, then remove that hack once we've run a data migration? Otherwise there could be a race condition: a new project (call it project XYZ) is created after I run a data migration, but before I deploy this code. So project XYZ's config is created without a field label for `senses` (because that default field label change in LexConfiguration is part of this PR), and so that project (and only that project) ends up getting "senses" in the activity log instead of "Meaning".

This would be a bad idea if we had implemented UI translations at the data level (e.g., if a Thai-language project had the string ความหมาย in Mongo as the label for the Senses field). But since we currently only do that UI translation at the presentation level (it's still "Meaning" in MongoDB) then that hack wouldn't cause any ill effects as long as I removed it promptly.

I'll write the hack but push it as a separate commit so that it's trivial to revert if you're opposed to the idea.
</blockquote></details>

The hack in question is commit 360b4e1e48fb416ab586b087f8c4a22a1fd97f25.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/265)*

**Comment by megahirt** [comment]
The assumption is that the migration is run as close to the deployment as possible.

---

Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-:-LDb8uRG9cqSLcqatRtF:b-l3zjbu)*

**Comment by irahopkinson** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/Api/Library/Shared/Script/Migration/FixSenseAndExampleLabels.php, line 83 at r2](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDcjHP_AD-_6HdBBjYH:-LDcjHPbB11CxUbTYvB7:b-eway3i) ([raw file](https://github.com/sillsdev/web-languageforge/blob/360b4e1e48fb416ab586b087f8c4a22a1fd97f25/src/Api/Library/Shared/Script/Migration/FixSenseAndExampleLabels.php#L83)):*
> ```
> FixSenseAndExampleLabels::run('test');
> ```
Changing to below would allow you to change the mode at run time without editing the file.
```
$mode = 'test';
if (isset($argv[1])) {
    $mode = $argv[1];
}
print "Running in $mode mode\n";
FixSenseAndExampleLabels::run($mode);
```

---

*[src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php, line 56 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJoFVxD6IShRuuYTG2:-LDcl9vz6XQjP4dRu9wH:b-s3v0b7) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php#L56)):*
<details><summary><i>Previously, rmunn (Robin Munn) wrote…</i></summary><blockquote>

The hack in question is commit 360b4e1e48fb416ab586b087f8c4a22a1fd97f25.
</blockquote></details>

I'm not in favour of that hack as its in production code and requires a second deploy to live to fix. Instead, in this case the migration script can be safely run both before and **after** the code is shipped to live.

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/265)*

**Comment by rmunn** [comment]
Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/Api/Library/Shared/Script/Migration/FixSenseAndExampleLabels.php, line 83 at r2](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDcjHP_AD-_6HdBBjYH:-LDkIBlnD6MBrWGcEiwn:b-896fix) ([raw file](https://github.com/sillsdev/web-languageforge/blob/360b4e1e48fb416ab586b087f8c4a22a1fd97f25/src/Api/Library/Shared/Script/Migration/FixSenseAndExampleLabels.php#L83)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

> ```
> FixSenseAndExampleLabels::run('test');
> ```
Changing to below would allow you to change the mode at run time without editing the file.
```
$mode = 'test';
if (isset($argv[1])) {
    $mode = $argv[1];
}
print "Running in $mode mode\n";
FixSenseAndExampleLabels::run($mode);
```
</blockquote></details>

Done.

---

*[src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php, line 56 at r1](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-LDJoFVxD6IShRuuYTG2:-LDkHwYu-I-hQq9GaKLq:bkxh6vn) ([raw file](https://github.com/sillsdev/web-languageforge/blob/c3993f8e7556180131ffcb97e02350816c0970b8/src/Api/Model/Languageforge/Lexicon/Command/LexEntryCommands.php#L56)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

I'm not in favour of that hack as its in production code and requires a second deploy to live to fix. Instead, in this case the migration script can be safely run both before and **after** the code is shipped to live.
</blockquote></details>

Reverted the hack; will run migration before shipping to live instead (by manually scp'ing the migration script to live and then running it).

---


*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/265)*

**Comment by irahopkinson** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/sillsdev/web-languageforge/265#-:-LDmxLWV1iSS7He0fpfS:bnfp4nl)*
